pkg/obfuscate/shell.go

```diff
@@ -0,0 +1,607 @@
+				nbrGrabbed++
+			}
+
+			break
+		} else if kind == SingleQuote || kind == DoubleQuote || kind == Backticks {
+			nbrGrabbed++
+			index++
+
+			// Grab the whole quoted string
+			for ; index < tokensLength && tokens[index].Kind != kind; index++ {
+				nbrGrabbed++
+			}
+
+			// Add the last token if it is the same as the start token
+			if index < tokensLength && tokens[index].Kind == kind {
+				nbrGrabbed++
+			}
+
+			break
+		} else if kind == Executable {
+			nbrGrabbed++
+			break
+		} else {
+			// Not a known token kind
+			break
+		}
+	}
+
+	if index >= tokensLength {
+		index--
+	}
+
+	return index, nbrGrabbed
+}
+
+// obfuscateCommandTokens obfuscates the given command arguments as a ShellToken array.
+// isExecCmd is true if the command is an exec command and preObfuscatedIndex is the index of the shell command for exec commands
+// returns the ShellToken array with obfuscated tokens and the indices of tokens that were obfuscated as a string array
+func obfuscateCommandTokens(tokens []ShellToken, isExecCmd bool, preObfuscatedIndex int) []string {
+	var obfuscatedArgsIndices []string
+
+	foundBinary := false
+	skipNext := false
+
+	for index := 0; index < len(tokens); index++ {
+		token := tokens[index]
+
+		if skipNext {
+			skipNext = false
+			continue
+		}
+
+		// Skip the obfuscation of the shell command for exec commands (already obfuscated)
+		if index == preObfuscatedIndex {
+			continue
+		}
+
+		if !foundBinary {
+			// Is this an environment variable? Assume that the format match our regex
+			if result := regexEnvVars.FindStringSubmatch(token.Val); result != nil {
+				// If this is an environment variable, check if it’s part of our passlist
+				if !stringInSlice(token.Val, allowedEnvVars) {
+					tokens[index].Val = result[1] + "=?"
+					obfuscatedArgsIndices = append(obfuscatedArgsIndices, strconv.Itoa(index)+"-"+strconv.Itoa(len(result[1])+1))
+				}
+			} else {
+				// If not formatted like an env variable, likely the binary
+				if stringInSlice(token.Val, processDenyList) {
+					// Remove every parameter until the end of the command
+					if isExecCmd {
+						// Remove all tokens until the end (the whole command)
+						for index++; index < len(tokens); index++ {
+							tokens[index].Val = "?"
+							obfuscatedArgsIndices = append(obfuscatedArgsIndices, strconv.Itoa(index))
+						}
+					} else {
+						for index++; index < len(tokens) && tokens[index].Kind == Field; index++ {
+							tokens[index].Val = "?"
+							obfuscatedArgsIndices = append(obfuscatedArgsIndices, strconv.Itoa(index))
+						}
+					}
+				}
+
+				foundBinary = true
+			}
+		} else { // Alright, we’re in the parameters then
+			// Are we in the case of --pass=xxx or --pass xxx
+			if equalIndex := strings.Index(token.Val, "="); equalIndex == -1 {
+
+				// if --pass xxx, check is --pass is allowed and that we have a xxx
+				if regexParamDeny.MatchString(token.Val) && index < len(tokens)-1 {
```

Comment:
```suggestion
				if regexParamDeny.MatchString(token.Val) && index+1 < len(tokens) {
```
To be consistent with previous comparisons (unless there's a reason it's flipped?)